### PR TITLE
Enable Limiter nesting with AllOf, AnyOf, NoneOf

### PR DESCRIPTION
**Pull request by henningkoehlernz**
Samples of new format (do not inherit to undead or devils):

	"limiters" : [
		"noneOf",
		"IS_UNDEAD",
		{
			"type" : "CREATURE_TYPE_LIMITER",
			"parameters" : [ "devil", true ]
		}
	]

Or the logical equivalent (to show off nesting)

	"limiters" : [
		"allOf",
		[
			"noneOf",
			"IS_UNDEAD"
		],
		[
			"noneOf",
			{
				"type" : "CREATURE_TYPE_LIMITER",
				"parameters" : [ "devil", true ]
			}
		]
	]

Or in compact format (allOf prefix is optional)

	"limiters" : [
		[ "noneOf", "IS_UNDEAD" ],
		[ "noneOf",
			{
				"type" : "CREATURE_TYPE_LIMITER",
				"parameters" : [ "devil", true ]
			}
		]
	]

### CODE REVIEW

File: lib/HeroBonus.cpp [hunk 1]
```diff
@@ -1840,7 +1842,49 @@ int LimiterList::limit( const BonusLimitationContext &context ) const
 	return wasntSure ? ILimiter::NOT_SURE : ILimiter::ACCEPT;
 }
 
-void LimiterList::add( TLimiterPtr limiter )
+void AllOfLimiter::add(TLimiterPtr limiter)
+{
+	limiters.push_back(limiter);
+}
+
+int AnyOfLimiter::limit(const BonusLimitationContext & context) const
```
Comment (alexvins) [inline lib/HeroBonus.cpp]: All that `limit` functions duplicate each other at 99% and (important) differences are not that easy to see. You should merge "logical" limiters into one class

File: lib/JsonNode.cpp [hunk 2]
```diff
@@ -567,6 +567,96 @@ void JsonUtils::resolveIdentifier(const JsonNode &node, si32 &var)
...
+	case JsonNode::JsonType::DATA_STRING: //pre-defined limiters
+		return parseByMap(bonusLimiterMap, &limiter, "limiter type ");
+		break;
+	case JsonNode::JsonType::DATA_STRUCT: //customizable limiters
+		{
+			std::string limiterType = limiter["type"].String();
+			const JsonVector & parameters = limiter["parameters"].Vector();
+			if(limiterType == "CREATURE_TYPE_LIMITER")
+			{
+				std::shared_ptr<CCreatureTypeLimiter> creatureLimiter = std::make_shared<CCreatureTypeLimiter>();
```
Comment (ArseniyShestakov) [inline lib/JsonNode.cpp]: I remember that you dislike using ```auto``` and I agree it make sense in some parts of code where readability is more important. Yet I don't think object creation like this is one of these cases.

Comment (henningkoehlernz) [reply lib/JsonNode.cpp]: I agree it's not needed here, but I don't see it causing any harm either.

**Comment by ArseniyShestakov** [comment]
Yes please add some note into change log. This change is important enough to mention it there.

File: lib/JsonNode.cpp (hunk 2 above)
Comment (ArseniyShestakov) [reply lib/JsonNode.cpp]: It's not really important. I just seen there is ```auto``` in other methods so I just noticed this one.

File: lib/registerTypes/RegisterTypes.h [hunk 3]
```diff
@@ -139,6 +139,9 @@ void registerTypesMapObjectTypes(Serializer &s)
 	s.template registerType<IUpdater, GrowsWithLevelUpdater>();
 	s.template registerType<IUpdater, TimesHeroLevelUpdater>();
 	s.template registerType<IUpdater, TimesStackLevelUpdater>();
+
+	s.template registerType<ILimiter, AnyOfLimiter>();
+	s.template registerType<ILimiter, NoneOfLimiter>();
```
Comment (alexvins) [inline lib/registerTypes/RegisterTypes.h]: You should adjust serialize format version when changing type registration

**Pull request merged**